src/interop/js-interop/usage.md

```diff
@@ -0,0 +1,368 @@
+JS type and just *viewing* it differently, so there is no conflict. If we
+wanted, we could have made the representation type `JSArray`:
+
+```dart
+extension type Array._(JSArray<JSAny?> _) implements JSArray<JSAny?> {}
+```
+
+Lastly, like any other extension type, you're allowed to declare any
+non-`external` members in the interop type. `isZeroLength` is one such example.
+
+#### Object literal constructors
+
+It is useful sometimes to create a JS [object literal] that simply contains a
+a number of properties and their values. In order to do this, we allow you to
+use a constructor with only named arguments:
+
+```dart
+import ‘dart:js_interop’;
+
+extension type Options._(JSObject o) {
+  external Options({int a, int b});
+  external int get a;
+  external int get b;
+}
+```
+
+A call to `Options(a: 0, b: 1)` will result in a JS invocation of
+`{a: 0, b: 1}`. You can get or set these values through `external` instance
+members.
+
+### Extension members on interop types
+
+You can also write `external` members in extensions of interop types. For
+example:
+
+```dart
+extension on Array {
+  external int push(JSAny? any);
+}
+```
+
+The semantics of calling `push` are identical to what it would have been if it
+was in the definition of `Array` instead. The only `external` members you can
+write in such an extension are instance members and the allowed operators. Like
+with interop types, you can write any non-`external` members in the extension.
+These extensions are useful for when an interop type doesn't expose the
+`external` member you need and you don't want to create a new interop type.
+
+### Parameters
+
+`external` interop methods can only contain positional and optional arguments.
+The one exception is object literal constructors, where they can contain *only*
+named arguments.
+
+Unlike with non-`external` methods, optional arguments do not get replaced with
+their default value, but are instead omitted. For example:
+
+```dart
+external int push(JSAny? any, [JSAny? any2]);
+```
+
+Calling `array.push(0.toJS)` in Dart will result in a JS invocation of
+`array.push(0.toJS)` and *not* `array.push(0.toJS, null)`. If you declare a
+parameter with an explicit default value, you will get a warning that the value
+will be ignored.
+
+## `@JS()`
+
+It is sometimes useful to refer to a JS property with a different name than the
+one written. For example, if you want to write two `external` APIs that point to
+the same JS property, you’d need to write a different name for at least one of
+them. In order to do this, you can use the [`@JS()`] annotation with a constant
+string value. For example:
+
+```dart
+extension type Array._(JSArray<JSAny?> _) implements JSArray<JSAny?> {
+  external int push(JSNumber number);
+  @JS('push')
+  external int pushString(JSString string);
+}
+```
+
+Calling either `push` or `pushString` will result in JS code that uses `push`.
+
+You can also rename interop types:
```

Comment:
Is it the case that without `@JS()` you _must_ use the name of the native JS type you're writing an extension type over as the name of the extension type? It makes sense that you'd have to do this to interact with APIs since you're calling `external` on all of them, but I thought maybe the interop types could be named anything. Like, in [the `Array` example](https://github.com/dart-lang/site-www/pull/4578/files#diff-173da65856efe1cfdb286777d80c04a313dad6171f415b1403693da8f8a95e29R73) up above, does that extension type _need_ to be named `Array`?  

If that's the case, and if it's not extremely obvious enough already, maybe you could add that to a sentence after [line 92](https://github.com/dart-lang/site-www/pull/4578/files#diff-173da65856efe1cfdb286777d80c04a313dad6171f415b1403693da8f8a95e29R92):

```md
Here we are using Array to interop with JS’ [Array] type. 
The extension type must have the same name as the native JS type
(unless using the `@JS` annotation (link to this section)).
```